runtime/bigarray.c

```diff
@@ -99,16 +99,15 @@ caml_ba_alloc(int flags, int num_dims, void * data, intnat * dim)
   for (i = 0; i < num_dims; i++) dimcopy[i] = dim[i];
   size = 0;
   if (data == NULL) {
+    elt_size = caml_ba_element_size[flags & CAML_BA_KIND_MASK];
     num_elts = 1;
     for (i = 0; i < num_dims; i++) {
       if (caml_umul_overflow(num_elts, dimcopy[i], &num_elts))
         caml_raise_out_of_memory();
     }
-    if (caml_umul_overflow(num_elts,
-                           caml_ba_element_size[flags & CAML_BA_KIND_MASK],
-                           &size))
+    if (caml_umul_overflow(num_elts, elt_size, &size))
       caml_raise_out_of_memory();
-    data = malloc(size);
+    data = calloc(num_elts, elt_size);
     if (data == NULL && size != 0) caml_raise_out_of_memory();
```

Comment:
This is approaching code golf-level pedantic C philosophy hour, but I wonder if technically speaking a `calloc` implementation could decide to always return a pointer for all calls, even empty ones, say for profiling/debugging purposes. And to update its table of allocated pointers it could require space that was not available, meaning the zero-sized allocation was unsuccessful, and thus it would return NULL and set `errno = ENOMEM` in order to signal an OOM situation to the application. But since it doesn't clear `errno` in the successful case we'd have no way of knowing without setting errno prior to calling `calloc`? Anyway, not relevant for this PR :)

Reply:
Both malloc and calloc can return a non-null ptr when size = 0.
https://linux.die.net/man/3/malloc

Even if the zero-sized allocation is unsuccessful and it returns NULL, there are no side effects that actually matter.